src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -44,51 +44,52 @@ public bool StartAssignment(HostAssignmentContext assignmentContext)
                 {
                     if (_assignmentContext != null)
                     {
-                        return _assignmentContext.Equals(assignmentContext);
+                        return _assignmentContext.Equals(context);
                     }
-                    _assignmentContext = assignmentContext;
+                    _assignmentContext = context;
                 }
 
-                // Queue the task in background, and return to caller immediately
-                Task.Run(() => Specialize(assignmentContext));
+                await Specialize(context);
 
                 return true;
             }
             else
             {
                 // No lock needed here since _assignmentContext is not null when we are here
-                return _assignmentContext.Equals(assignmentContext);
+                return _assignmentContext.Equals(context);
             }
         }
 
         private async Task Specialize(HostAssignmentContext assignmentContext)
         {
             try
             {
-                var zip = assignmentContext.ZipUrl;
-                if (!string.IsNullOrEmpty(zip))
-                {
-                    // download zip and extract
-                    var filePath = Path.GetTempFileName();
-
-                    await DownloadAsync(new Uri(zip), filePath);
-
-                    assignmentContext.ApplyAppSettings();
-
-                    ZipFile.ExtractToDirectory(filePath, _webHostSettings.ScriptPath, overwriteFiles: true);
-                }
-                else
-                {
-                    assignmentContext.ApplyAppSettings();
-                }
+                // first make all environment and file system changes required for
+                // the host to be specialized
+                await ApplyContext(assignmentContext);
 
-                // set flags which will trigger specialization
+                // all assignment settings/files have been applied so we can flip
+                // the switch now on specialization
                 _settingsManager.SetSetting(EnvironmentSettingNames.AzureWebsitePlaceholderMode, "0");
```

Comment:
The crucial change here is that we don't return back to the caller of Assign until we've flipped these environment flags for specialization. That ensures that a subsequent http request will initiate specialization, and that request will block until the host is specialized and the function can be processed.

Previously we were just doing a Task.Run, meaning a subsequent http request could come in for example while we were unpacking the zip and before we flipped the bits, resulting in a 404.

Reply:
By 'Azure request timeout', do you mean the 4 minute load balancer timeout? If it takes longer than that, then I'm don't think there is anything we can do to honor the cold start request, since that request can't be served till this is complete. So I don't think waiting is wrong here. Hopefully, it takes an extreme scenario to break this.